### DISCUSSION TITLE
Issue importing iText .net lib using pythonnet

### DISCUSSION THREAD

**Discussion by typhoon71** [pythonnet/pythonnet#1688] [Q&A]
### Environment

-  Pythonnet version: 2.5.1
-  Python version: 3.7
-  Operating System: Windows7
- iText version 7.1.11.0 (dotnet4)
### Details
I'm using Python.NET to create small wrapper for iText (https://github.com/itext/itext7-dotnet).
With Jetbrain dotPeek I see that itext.kernel (the dll I'm trying to use) has various namespaces; the namespace iText.Kernel.Pdf is what I need, in particular the public class PdfReader is what I want to use.

Here's what I'm doing:
```python
import clr
clr.AddReference('itext.kernel')
sys.path.append(os.getcwd())  # the script is currently in the dll folder
from iText.Kernel.Pdf import PdfReader
```
But the last line gets me a:
```python
 "ModuleNotFoundError: No module named 'iText'" error.
```
I am able to load other dlls and use them just fine, but I don't get what is wrong here.
Can someone please point me to the right direction? Thanks.

edit: I can import and use iTextSharp just fine (different namespaces / classes), it seems iText7 has something different.

**Comment by typhoon71** [comment]
Well, I'm answering my own question, it may help someone else.

I found out what the issue was and was able to make it work. Here's the details.

First thing I did was testing the other iText libs:
```python
import clr
clr.AddReference('itext.forms')
from iText.Forms import PdfAcroForm
```
This worked flawlessy, so the issue must've been in `itext.kernel`.

I tried using older builds, and I found out that build 7.0.8 worked fine.

I compared it with build 7.1.11 (the one I was having problems with) using dotPeek, and noticed it has some references, namely  `BouncyCastle.Crypto`, `Common.Logging` and `Common.Logging.Core`.

I did put them in the project folder, but still no dice... until I noticed I was using `BouncyCastle.Crypto` build 1.8.6, while the `itext.kernel` referred build 1.8.5! Replacing with that version solved the issue.

_Basically the newer builds of iText7 have 3 referenced libs, that must be present in the exact same version._

**THE QUESTION**

Would it be possible for pythonnet to be more verbose about this?

The `"ModuleNotFoundError: No module named 'iText'"` error wasn't helpful at all, since the `clr.AddReference` didn't hint at any problem and I'd expet some exception to be trown here (MissingReferenceException? ReferenceVersionException?).

I'm just asking, I don't know if it's feasible. Also, thanks a lot for pythonnet! Just found it and already find it useful.

**Comment by retsyo** [comment]
Thank you, I met `ModuleNotFoundError: No module named 'iText'` when [itext7.7.1.17](https://github.com/itext/itext7-dotnet/releases/tag/7.1.17) is used. After I add `BouncyCastle.Crypto.dll`, `Common.Logging.Core.dll` and `Common.Logging.dll`, it works.

however, for [itext7.7.2.0](https://github.com/itext/itext7-dotnet/releases/tag/7.2.0), even `itext.commons.dll`, `BouncyCastle.Crypto.dll` and `Microsoft.Extensions.Logging.Abstractions.dll` are added, `ModuleNotFoundError: No module named 'iText'` still appears. Maybe there are extra attention I have to pay.

**Comment by typhoon71** [comment]
correct: for iText7 7.2.2.0 you need Microsoft.Extensions.Logging.Abstractions.dll (v5.0.0), since the Microsoft.Extensions.Logging framework was introduced as the new iText code logging module

This change adds the need for Microsoft.Extensions.Logging.dll and Microsoft.Extensions.Options.dll too, as you can see by some use of dotPeek on itext.commons.dll and the error messages.

Add those 2 and it should run just fine (at least my code does).

**Comment by retsyo** [comment]
now it works, thank you. I forgot to check `itext.commons.dll`
btw, is there any tools/code which can check and report on one page about this kind of hierarchical dependency missing?

**Comment by typhoon71** [comment]
You can select "Inspect", "Reference Hierarchy" and see the whole tree of dependencies referenced.

Or you could try ILSpy (7.2 Preview 3) which shows all references of a dll, recursively, not just the top level ones, in the first tree - seems faster to me that dotPeek, may be a viable alternative.

**Comment by retsyo** [comment]
the commandline tool [asmspy](https://github.com/mikehadlow/AsmSpy) can show the dependencies. 

the red text means something is wrong. However, it also displays redundant missing

```
cd where_C#_DLLs_live
asmspy .
```
